compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAcquireReleaseToUseLock.cpp

```diff
@@ -0,0 +1,234 @@
+      if (!accessOp) {
+        acqOp.emitOpError() << "currently only supports "
+                               "`amdaie.logicalobjectfifo.access` users";
+        return WalkResult::interrupt();
+      }
+      AMDAIE::BufferOp bufferOp = buffers[bufferIndex];
+      accessOp.getResult().replaceAllUsesWith(bufferOp.getResult());
+      toBeErased.push_back(accessOp);
+    }
+    logicalObjFifoToIndex[logicalObjFifo] += acqSize;
+    toBeErased.push_back(acqOp);
+    return WalkResult::advance();
+  });
+  if (res.wasInterrupted()) return failure();
+  for (Operation *op : toBeErased) {
+    op->dropAllUses();
+    rewriter.eraseOp(op);
+  }
+  return success();
+}
+
+LogicalResult releaseToUseLock(RewriterBase &rewriter, AMDAIE::CoreOp coreOp) {
+  OpBuilder::InsertionGuard g(rewriter);
+  AMDAIE::TileOp tileOp = coreOp.getTileOp();
+  SmallVector<Operation *> toBeErased;
+  WalkResult res = coreOp.walk([&](AMDAIE::LogicalObjectFifoRelease relOp) {
+    LLVM_DEBUG(llvm::dbgs() << "Convert release op: " << relOp << "\n");
+    std::optional<int> maybeRelSize = relOp.getSize();
+    assert(maybeRelSize && maybeRelSize.value() == 1 &&
+           "logic currently only handles size set and equal to 1");
+    int relSize = maybeRelSize.value();
+
+    FailureOr<AMDAIE::LogicalObjectFifoFromBuffersOp> maybeLogicalObjFifo =
+        getLogicalObjFifoOperatedOn<AMDAIE::LogicalObjectFifoRelease>(relOp);
+    if (failed(maybeLogicalObjFifo)) return WalkResult::interrupt();
+
+    FailureOr<AMDAIE::LockOp> maybeLockOp =
+        getLockToBeUsed(maybeLogicalObjFifo.value(), tileOp, relOp.getPort(),
+                        LockAction::Release);
+    if (failed(maybeLockOp)) return WalkResult::interrupt();
+
+    rewriter.setInsertionPoint(relOp);
+    rewriter.create<AMDAIE::UseLockOp>(relOp.getLoc(), maybeLockOp.value(),
+                                       LockAction::Release, relSize);
+    toBeErased.push_back(relOp);
+    return WalkResult::advance();
+  });
+  if (res.wasInterrupted()) return failure();
+  for (Operation *op : toBeErased) {
+    op->dropAllUses();
+    rewriter.eraseOp(op);
+  }
+  return success();
+}
+
+namespace {
+
+struct AMDAIEAcquireReleaseToUseLockPass
+    : public impl::AMDAIEAcquireReleaseToUseLockBase<
+          AMDAIEAcquireReleaseToUseLockPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+
+  void runOnOperation() override {
+    Operation *parentOp = getOperation();
+    IRRewriter rewriter(parentOp->getContext());
+
+    WalkResult res = parentOp->walk([&](AMDAIE::CoreOp coreOp) {
+      // Loops need to be unrolled based on on the depths of the logical
+      // objectFifos so `amdaie.use_lock` ops can be inserted correctly for
+      // double buffering purposes, without need for a dependency on the loop
+      // induction variable.
+      if (failed(coreLoopUnroll(rewriter, coreOp))) {
+        return WalkResult::interrupt();
+      }
```

Comment:
Why don't we separate this out of this pass ?

Reply:
This unrolling is very specific to the conversion from acquire/release to use_lock, which would not be valid without the unrolling first.